src/common/lua/ngx_wasm_lua_ffi.c

```diff
@@ -256,4 +257,173 @@ ngx_http_wasm_ffi_set_host_properties_handlers(ngx_http_request_t *r,
 
     return ngx_proxy_wasm_properties_set_ffi_handlers(pwctx, getter, setter, r);
 }
+
+
+ngx_int_t
+ngx_wa_ffi_shm_get_zones(ngx_wa_ffi_shm_get_zones_handler_pt handler)
+{
+    ngx_uint_t             i;
+    ngx_array_t           *shms = ngx_wasmx_shms((ngx_cycle_t *) ngx_cycle);
+    ngx_wa_shm_t          *shm;
+    ngx_wa_shm_mapping_t  *mappings;
+
+    if (!handler || !shms) {
+        return NGX_ABORT;
+    }
+
+    mappings = shms->elts;
+
+    for (i = 0; i < shms->nelts; i++) {
+        shm = mappings[i].zone->data;
+        handler(shm);
+    }
+
+    return NGX_OK;
+}
+
+
+static void
+shm_kv_retrieve_keys(ngx_rbtree_node_t *node, ngx_rbtree_node_t *sentinel,
+                     ngx_int_t *total, ngx_int_t limit, ngx_str_t **keys)
+{
+    ngx_wa_shm_kv_node_t *n = (ngx_wa_shm_kv_node_t *) node;
+
+    if (!node || node == sentinel || (limit > 0 && *total == limit)) {
+        return;
+    }
+
+    if (keys) {
+        keys[*total] = &n->key.str;
+    }
+
+    (*total)++;
+
+    shm_kv_retrieve_keys(node->right, sentinel, total, limit, keys);
+    shm_kv_retrieve_keys(node->left, sentinel, total, limit, keys);
+}
+
+
+ngx_int_t
+ngx_wa_ffi_shm_get_keys(ngx_wa_shm_t *shm, ngx_uint_t n, ngx_str_t **keys)
+{
+    ngx_int_t         total = 0;
+    ngx_wa_shm_kv_t  *kv;
+
+    if (!shm || shm->type == NGX_WASM_SHM_TYPE_QUEUE) {
```

Comment:
`shm == NULL` (ditto for all NULL comparison with `!` in the PR - `!` is only for numbers in our convention).